ch_pipeline/processing/daily.py

```diff
@@ -406,21 +408,103 @@ def _load_hook(self):
         # Process the intervals
         self._intervals = []
         for t in self._revparams["intervals"]:
-            self._intervals.append((t["start"], t.get("end", None)))
+            self._intervals.append((t["start"], t.get("end", None), t.get("step", 1)))
+
         self._padding = self._revparams["padding"]
 
+    def _available_files(self, start_csd, end_csd):
+        """
+        Return chimestack files available in cedar_online between start_csd and end_csd, if all of the files for that period are available online.
+
+        Return an empty list if files between start_csd and end_csd are only partially available online.
+
+        Total file count is verified by checking files that exist everywhere.
+
+        Parameters
+        ----------
+        start_csd : int
+            Start date in sidereal day format
+        end_csd : int
+            End date in sidereal day format
+
+        Returns
+        -------
+        list
+            List contains the chimestack files available in the timespan, if all of them are available online
+
+        """
+
+        # Connect to databases
+        db.connect()
+
+        # Get timestamps in unix format
+        # Needed for queries
+        start_time = ephemeris.csd_to_unix(start_csd)
+        end_time = ephemeris.csd_to_unix(end_csd)
+
+        # We will want to know which files are in chime_online and nearline on cedar
+        online_node = di.StorageNode.get(name="cedar_online", active=True)
+        chimestack_inst = di.ArchiveInst.get(name="chimestack")
+
+        # TODO if the time range is so small that it’s completely contained within a single file, nothing will be returned
+        # have to special-case it by looking for files which start before the start time and end after the end time).
+
+        archive_files = (
+            di.ArchiveFileCopy.select(
+                di.ArchiveFileCopy.node,
+                di.ArchiveAcq.name.alias("acq_name"),
+                di.ArchiveFile.name.alias("file_name"),
+                di.CorrFileInfo.start_time,
+                di.CorrFileInfo.finish_time,
+            )
+            .join(di.ArchiveFile)
+            .join(di.ArchiveAcq)
+            .switch(di.ArchiveFile)
+            .join(di.CorrFileInfo)
+        )
+
+        # chimestack files available online which include between start and end_time
+
+        files_that_exist = archive_files.where(
+            di.ArchiveAcq.inst
+            == chimestack_inst,  # specifically looking for chimestack files
+            di.CorrFileInfo.start_time
+            < end_time,  # which contain data that includes start time and end time
+            di.CorrFileInfo.finish_time >= start_time,
+            di.ArchiveFileCopy.has_file == "Y",
+        )
+
+        files_online = files_that_exist.where(
+            di.ArchiveFileCopy.node == online_node,  # that are online
+        )
+
+        ChimestackFile = namedtuple("ChimestackFile", "name start_time finish_time")
```

Comment:
I think my comment about about peewee objects vs tuples vs named tuples was a bit unclear, but hopefully this makes more sense...

Using peewee objects gives you a nice well defined interface to the query results, with named attributes for accessing the information. However, you pay a bit of a price in object construction.

Using bare tuples is simpler and more efficient, at the price of there being no standard interface, and you can't use the returned tuples for the usual peewee magic queries (as they're not fancy objects). But if you know you've extracted all the information you need in that query then this is fine. I think you are in this regime here as these queries return all the info you need.

The named tuple approach here tries to combine both, but I think doesn't really bring through the advantages of either. It lacks the standard and expected interface defined by the data_index tables that the first approach would give, and it requires an extra conversion to a namedtuple in there which probably negates some of the efficiency and simplicity of the second approach.

I guess my recommendation for the future is to use one of the first two approaches. If you think it's reasonable to just use plain old tuples then do so, otherwise use peewee's objects directly. It's definitely not worth changing this now.

Reply:
> standard interface

This is a very good point against Namedtuples!

Reply:
We ended up switching to tuples!